### PR TITLE
FixedChannelPool acquired connection counter bug fixed

### PR DESCRIPTION
**Pull request by ioanbsu**
Motivation:
The FixedChannelPool once requested for a channel from the pool that is unhealthy will:
a) close this channel
b) will try to acquire new channel if acquiredChannelCount < maxConnections.
c) If acquiredChannelCount >= maxConnections it will add task for the queue.
Now for the case with maxConnections==1 this will cause infinite grow of pendingAcquireQueue in FixedChannelPool because acquire is not decremented after step a) when SimpleChannelPool realised that channel is unhealthy.
For mode details see: #4023

Modifications:
Added promises to notifyHealthCheck function so that when unhealthy channel closing routine is done it will call unAcquireChannel function that is overridden in FixedChannelPool and inside thins function it will call decrementAndRunTaskQueue(). This way we will make sure that when unhealthy channel pulled from queue after it closed we decrement the acquiredChannelCount as well as we force to check if there any tasks in queue that may be executed.

Result:
The bug where acquiredChannelsCount calculated incorrectly had been fixed. Now when pulled channel from queue turned out to be unhealthy the acquired channels count is decremented.

P.S. this is copy of https://github.com/netty/netty/pull/4027. Sorry for the confusion. I thought I was suppose to squash all my commits, but that seemed to just close my original pull request.

### CODE REVIEW

**Comment by ioanbsu** [comment]
@normanmaurer I applied changes that you asked, sorry for the confusion. Seems the fact that I did force push(git push origin +fixed_channel_pool_counter_fix ) close my previous pull request: https://github.com/netty/netty/pull/4027

**Comment by netkins** [comment]
**AUTOMATED MESSAGE for ADMINS**:<br>Please verify and accept the pull request.<br>To accept, say `@netkins accept`<br>To build again, say `@netkins build`<br>To whitelist the author, say `@netkins whitelist`

**Comment by normanmaurer** [comment]
Did you sign the icla yet : http://netty.io/s/icla ?

**Comment by ioanbsu** [comment]
Yes I did.

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 1]
```diff
@@ -161,19 +169,32 @@ private void notifyHealthCheck(Future<Boolean> future, Channel ch, Promise<Chann
                     handler.channelAcquired(ch);
                     promise.setSuccess(ch);
                 } catch (Throwable cause) {
-                    closeAndFail(ch, cause, promise);
+                    Promise<Void> channelClosePromise = ch.eventLoop().newPromise();
+                    channelClosePromise.addListener(channelWasNotAcuired);
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: channelWasNotAcuired -> channelWasNotAcquired (typo)

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 2]
```diff
@@ -50,6 +51,13 @@
     private final ChannelPoolHandler handler;
     private final ChannelHealthChecker healthCheck;
     private final Bootstrap bootstrap;
+    private GenericFutureListener<Future<? super Void>> channelWasNotAcuired =
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: final

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 3]
```diff
@@ -218,28 +239,39 @@ private void doReleaseChannel(Channel channel, Promise<Void> promise) {
...
-    private static void closeChannel(Channel channel) {
+    private static void closeChannel(final Channel channel, final Promise channelClosePromise) {
         channel.attr(POOL_KEY).getAndSet(null);
-        channel.close();
+        ChannelFuture future = channel.close();
+        if (channelClosePromise != null) {
+            future.addListener(new GenericFutureListener<Future<? super Void>>() {
+                @Override
+                public void operationComplete(Future<? super Void> future) throws Exception {
+                    if (future.isSuccess()) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: Shouldn't we call also "channelClosePromise.setFailure(...)" if not success full ?

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java [hunk 4]
```diff
@@ -218,28 +239,39 @@ private void doReleaseChannel(Channel channel, Promise<Void> promise) {
...
         }
     }
 
-    private static void closeChannel(Channel channel) {
+    private static void closeChannel(final Channel channel, final Promise channelClosePromise) {
         channel.attr(POOL_KEY).getAndSet(null);
-        channel.close();
+        ChannelFuture future = channel.close();
+        if (channelClosePromise != null) {
+            future.addListener(new GenericFutureListener<Future<? super Void>>() {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: You can use FutureListener

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java (hunk 2 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: You can use FutureListener

**Comment by fratboy** [comment]
I don't understand this well. We call `super.acquire(task.promise)`, so if connection is unhealty, doesn't `task.promise` get called by `setFailure`?  If it does, `AcquireListener.operationComplete` will be called and `acquiredChannelCount` will be decreased.

**Comment by ioanbsu** [comment]
@fratboy [so if connection is unhealty, doesn't task.promise get called by setFailure?] - it does not. It just tries to reacquire new channel without decreasing ```acquiredChannelCount```. But I see that you suggesting that we may try to call original task.promise's setFailure.
The problem I see with it is that if I call task.promise.setFailure() it will trigger action to notify that channel failed to be acquired. So who ever uses channel pool will have to retry to acquire channel. Wether in my approach it will just remove unhealthy channel from the pool and will try to acquire new one.

**Comment by fratboy** [comment]
@ioanbsu You're right. Thanks for your kind explanation. I'll watch this issue.

File: transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java [hunk 5]
```diff
@@ -245,6 +245,15 @@ public void operationComplete(Future<Void> future) throws Exception {
         return p;
     }
 
+    /**
+     * Called after channel failed to be acquired so we know to decrease the {@link #acquiredChannelCount}
+     * and pull the next pending task from {@link #pendingAcquireQueue}.
+     */
+    @Override
+    protected void channelClosedCauseUnhealthy() {
+        decrementAndRunTaskQueue();
```
Comment (fratboy) [inline transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: Don't we have to execute this with `FixedChannelPool .executor` ?

File: transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java (hunk 3 above)
Comment (ioanbsu) [inline transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java]: Hm. should I just do
```java 
channelClosePromise.setFailure(new IllegalStateException("Failed to close unhealthy channel."));
```
?

**Comment by ioanbsu** [comment]
@normanmaurer I made al the changes that you requested(except for setFailure). but have not pushed them yet.
Question:
Should I squash my changes? if yes is there any instructions I should follow? Last time I tried to squash my changes it caused my original pull request be closed(probably cause I did force).

File: transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java (hunk 5 above)
Comment (ioanbsu) [inline transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java]: Sure.

**Comment by ioanbsu** [comment]
@normanmaurer, @fratboy  any updates on this? Am I required to make more changes in this pull request?

**Comment by fratboy** [comment]
@ioanbsu Please check my scenario is possible.

- `FixedChannelPool` try to acquire a channel by delegating to `SimpleChannelPool` -> `acquiredChannelCount = 1`
- `SimpleChannelPool` try to acquire a channel in queue but failed because it's unhealty. -> `channelClosedCauseUnhealthy` -> `acquiredChannelCount = 0`
- `SimpleChannelPool` try to make a channel (kind of recursive) but failed because it's not connectable -> `promise` passed by `FixedChannelPool` will be called -> `acquiredChannelCount = -1` (error)

**Comment by ioanbsu** [comment]
@fratboy  so in the last step you're describing, before trying to make a new channel (with ```acquire(promise)```) it will actually increase ```acquiredChannelCount``` by one. So it will become ```acquiredChannelCount = 1```
And then, if it's not connectable it will be closed and ```acquiredChannelCount``` will be decreased by one and become 0. Correct behavior in as far as I can see.

**Comment by fratboy** [comment]
@ioanbsu I omitted pre process.
- `FixedChannelPool` acquires an healty channel and uses and releases. -> `acquiredChannelCount = 1` -> `acquiredChannelCount=0` and `SimpleChannelPool` will have that channel in `SimpleChannelPool.deque`
- That channel is disconnected. (but still in `SimpleChannelPool.deque`)
- ...

**Comment by ioanbsu** [comment]
@fratboy I don't think I understand well the use case you describing, why ```acquiredChannelCount``` will become -1?
This is the call sequence: 
```SimpleChannelPool .acquire``` == ```FixhedChannelPool.acquire``` -> ```FixedChannelPool.acquire0``` -> [if there is space to acquire more channels it will ```++acquiredChannelCount```]-> super.acquire()==SimpleChannelPoo.acquire()


If FixedChannelPool is used it will increase ```acquiredChannelCount``` every single time it tries to acquire a channel. Later if it realizes that channel failed to be acquired for any reason it will decrement the ```acquiredChannelCount```. It will also decrement ```acquiredChannelCount``` when channel is released.

**Comment by fratboy** [comment]
@ioanbsu Here is then call sequence I thought.
- `FixedChannelPool.acquire` -> `FixedChannelPool.acquire0` -> `acquiredChannelCount++` -> `SimpleChannelPool.acquire(promise)`
- in `SimpleChannelPool.acquire` if `SimpleChannelPool.pollChannel()` is not null (one channel exists in dequeue) -> `SimpleChannelPool.doHealthCheck`
- in `SimpleChannelPool.doHealthCheck` -> `SimpleChannelPool.notifyHealthCheck` -> if a channel is not healty -> `SimpleChannelPool.closeChannel` -> `SimpleChannelPool.channelWasNotAcquired` -> `FixedChannelPool.channelClosedCauseUnhealthy` will be called -> `acquiredChannelCount--`
- Then `SimpleChannelPool.acquire` again -> `SimpleChannelPool.pollChannel()` will return null -> try to connect new channel -> if connection is unavailabe `SimpleChannelPool.notifyConnect()` will call `promise.setFailure` and it will call `acquiredChannelCount--`

**Comment by ioanbsu** [comment]
Got it.
 Ok so in your last statement [Then ```SimpleChannelPool.acquire``` again -> ```SimpleChannelPool.pollChannel()``` will return null -> try to connect new channel -> if connection is unavailabe ```SimpleChannelPool.notifyConnect()``` will call ```promise.setFailure``` and it will call ```acquiredChannelCount--```] SimpleChannelPool.acquire = FixedChannelPool.acquire cause acquire overrides SimpleChannelPool's acquire. 
So in reality your last step will look like:
```SimpleChannelPool.acquire``` won't be called right away, the ```FixedChannelPool.acquire``` which takes us to the very first step in list of steps that you just described.

**Comment by fratboy** [comment]
@ioanbsu You're right. I missed that point. But I think it still has a problem. 
Because in second try(call by `SimpleChannelPool.notifyConnect()`), `promise` is wrapped again before call `super.acquire()` in `FixedChannelPool.acquire0`. [FixedChannelPool.java#L199](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java#L199) So `acquiredChannelCount--` will be called twice.

**Comment by ioanbsu** [comment]
@fratboy,
To make sure I got you correctly let me describe step by step channel acquire process in details as I understand it.
So the use case you're proposing is

1. acquiredChannelCount = 0
```FixedChannelPool.acquire```->```FixedChannelPool.acquire0```-> [in ```FixedChannelPool.acquire0```] the [```acquiredChannelCount < maxConnections```]==true so we do ```acquiredChannelCount++```
2. acquiredChannelCount = 1. Create new promise(Let's call it promise1)->```SimpleChannelPool.acquire(promise1)```->in ```SimpleChannelPool.acquire(promise1)``` the ```pollChannel()``` returns non null value so we proceed to -> ```SimpleChannelPool.doHealthCheck(...,promise1)```-> ```SimpleChannelPool.notifyHealthCheck(promise1)```-> here the channel turned out to be unhealthy so following actions are performed:
 * create new ```channelClosePromise``` to listen for when channel was closed: ```Promise<Void> channelClosePromise = ch.eventLoop().newPromise();```
 * add listener to this ```channelClosePromise```: ```channelClosePromise.addListener(channelWasNotAcquired);```
 * close the unhealthy channel: ```closeChannel(ch, channelClosePromise);```
 * start acquiring new channel: ```acquire(promise1);```

3. The ```acquiredChannelCount``` is still =1, but as soon as channel will be closed the ```channelClosePromise``` will be notified and it will force the ```acquiredChannelCount--```, by calling ```FixedChannelPool.decrementAndRunTaskQueue()```
So ```acquiredChannelCount``` will become 0 again.
4. Looking further the ```acquire(promise1)``` had been called when channel was notified to be unhealthy.
```FixedChannelPool.acquire(promise1)```->```FixedChannelPool.acquire0(promise1)```->[acquiredChannelCount < maxConnections]==true->```++acquiredChannelCount```
5.```acquiredChannelCount==1```
create new promise(let's call it promise2) that all it will do is notify promise1 when action is triggered on promise2->```SimpleChannelPool.acquire(promise2)```->```pollChannel()==null``` so it will try to create a new channel-> new channel failed to be created will cause ```promise2.setFailure()```-> will only send signal to promise1 telling that channel was not acquired-> promise1 will send signal to AcquireListener that will call ```decrementAndRunTaskQueue()```->```--acquiredChannelCount```
6. acquiredChannelCount = 0.  Now ```decrementAndRunTaskQueue()``` may to start all over from step 1. just because there may be a task to acquire a channel is in pending queue, but this is totally fine and at the end the ```acquiredChannelCount``` will still be 0. 

As you can see in step 5. when it tries to create a new channel it does not even step into ```SimpleChannelPool.notifyHealthCheck(promise1)``` so the ```channelClosePromise``` will never be created second time and it won't force ```acquiredChannelCount``` be decremented twice. 

The only time when it steps into ```SimpleChannelPool.notifyHealthCheck(promise1)```  is when it pulled channel from the pool and realized that it was unhealthy. And only for that unhealthy channel the single promise ```channelClosePromise``` will be created so it won't be triggered twice even. Or if it will be triggered twice it will be doing it for two different channels that both turned out to be unhealthy. But since we call ```FixedChannelPool.acquire()``` every time the ```acquiredChannelCount``` will be also increased twice in this case and at the end when all channels are released and closed the ```acquiredChannelCount``` will be 0.

Please let me know what do you think?

**Comment by fratboy** [comment]
@ioanbsu in 4, `promise1` has `AcquireListener` that will decrease `acquiredChannelCount` when channel acquire fails. In second try `promise2` will have another `AcquireListener` that has `promise1` as `originalPromise`.  So when `promise2` is setFailure two `AcquireListener`s will be called in chain. 
How do u think?

**Comment by ioanbsu** [comment]
@fratboy why do you think that throwing ```setFailure``` on ```promise1``` will cause ```acquiredChannelCount``` be decreased?
There is no listeners who subscribed to promise1 or promise2 that will do ```acquiredChannelCount``` decrease.

**Comment by fratboy** [comment]
@ioanbsu Doesn't it?
```java
//first try
            Promise<Channel> p = executor.newPromise(); //p is promise1
            p.addListener(new AcquireListener(promise)); //promise is original promise user passed
            super.acquire(p); //p is promise1
//second time
            Promise<Channel> p = executor.newPromise(); //p is promise2
            p.addListener(new AcquireListener(promise)); //promise is promise1
            super.acquire(p); //p is promise2
```
[FixedChannelPool.java#L364](https://github.com/ioanbsu/netty/blob/fixed_channel_pool_counter_fix/transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java#L364)

**Comment by ioanbsu** [comment]
Ah I see what you saying.
Yep I was able to reproduce the behavior you describing.
Seems this is not the best idea to decrement the ```AcquireListener``` since there is possibility that setFailure may be ca,led several times. Even if we do correct math and make it end up ```acquiredChannelCount==0``` at the end I Don't think we should be ==-ing the ```acquiredChannelCount``` several times and then after several attempts do -- ```acquiredChannelCount``` several times.
I'd rather did ++ on the time when I attempted and did -- once right after I figured out the channel failed to be acquired.
Going to make some changes in my pull request to fix the issue that you describing.

**Comment by ioanbsu** [comment]
@fratboy push the code that fixed the use case you described.
P.S. Sorry I did no realise that the MD does not work in git when it reads commit messages.

**Comment by fratboy** [comment]
@ioanbsu If we don't call `decrementAndRunTaskQueue` in `ActionListener`, `acquiredChannelCount` will not decrease when `SimpleChannelPool.pollChannel() == null`(create a new channel) but `SimpleChannelPool.connectChannel(bs)` fail case. (Case when client tries to connect when server is down.)

I totally agree with u that we have to handle count of unhealthy connection from the pool, but I think we have to think about this case from scratch.

@normanmaurer What is your opinion?

**Comment by ioanbsu** [comment]
@fratboy  yes, you're right. So to solve this we just need to call ```channelClosedCauseUnhealthy()```  here https://github.com/ioanbsu/netty/blob/fixed_channel_pool_counter_fix/transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java#L142

In general I think we can not be using original promise that is passed recursively via recursive call of acquire, since this way we will end up adding (++) to```acquireChannelCount``` several times and then subtracting (--) several times at the end once we run out of channels from the pool and realize that new channel can not be created or when we finally acquire healthy channel from the pool. This is why I suggested that we need to have a function in ```SimpleChannelPool``` through what we can use to notify the ```FixedChannelPool``` to decrement the ```acquiredChannelCount``` so we can instantly decrement ```acquiredChannelCount``` and not wait when the original promise gets success of failure signal.

**Comment by ioanbsu** [comment]
@fratboy, @normanmaurer , thinking more about the issue with ```acquiredChannelCount``` I think I may be able to come up with the solution that does not needs ```FixedChannelPool.AcquireListener``` 
Would you like me to create another pull request from master where the ```acquiredChannelCount``` is calculated without ```AcquireListener```?

**Comment by ioanbsu** [comment]
@normanmaurer, @fratboy just created another pull request that does fix the same problem but much easier and simpler way:
https://github.com/netty/netty/pull/4076 it contains a test as well a test that fails to finish even without the changes suggested in pull request.

**Comment by normanmaurer** [comment]
superseded by #4076

**Pull request closed without merging**